README.md
# xhb_swiper

flutter 轮播图组件

![竖向滚动](https://media.giphy.com/media/vVgiJE5lhgeMpbLInK/giphy.gif)

![横向滚动](https://media.giphy.com/media/TQycRd0r0r1WW1QKwE/giphy.gif)
![demo](https://p1-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/88c08745ff524eebbd6dafc35247a66e~tplv-k3u1fbpfcp-watermark.image)

## 使用

1 Add this to your package's pubspec.yaml file:

```
dependencies:
  xhb_swiper: ^lastVersion
```

2 flutter pub get

3import 'package:xhb_swiper/xhb_swiper.dart';

## 提供的属性

![属性](https://p9-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/7d8a3b23658142a9833a76489839b7ea~tplv-k3u1fbpfcp-watermark.image)

## demo

```
// 自定指示器
  List<String> _listes = [
    "https://yy0.tv.cmvideo.cn:8443/migutv/res/2019/10/17/59L11DV8GKF0.jpg",
    "https://yy0.tv.cmvideo.cn:8443/migutv/res/2019/09/30/59IHTLCINHHH.jpg",
    "https://yy1.tv.cmvideo.cn:8443/migutv-clt/pomsimage/1504/026/730/202003240715_CCTVNEWS_1500000_20200324_35370063_0_327_HSJ1080H.jpg"
  ];

Swiper(
    _listes,
    onTap: (value) {
      print(value);
    },
    indicatorBuilder: (context, index) {
      return Container(
        width: 100,
        height: 30,
        alignment: Alignment.center,
        color: Colors.white,
        child: Text(
          '${index + 1}/3',
          style: TextStyle(
            color: Colors.red,
            fontWeight: FontWeight.w700,
            fontSize: 20,
          ),
        ),
      );
    },
  ),
```


CHANGELOG.md
## [0.0.5] - TODO: Add release date.

- TODO: Describe initial release.


test/xhb_swiper_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:xhb_swiper/xhb_swiper.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

// import 'package:swiper_demo/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    // await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


手写/web.js
//1********* 冒泡排序
/*
 冒泡就是两层循环，time:O(N^2)
思考：冒泡排序是稳定的吗？
解答：不稳定，因为出现了交换。
*/

const arr = [5,2,4,7,9,8,3,6,3,8,3,8];

const bubbleSort = (arr) => {
  for(let i=0; i<arr.length; i++){
    for(let j=0; j<arr.length - 1 - i; j++){
      if(arr[j] > arr[j+1]){
        let tmp = arr[j]
        arr[j] = arr[j+1]
        arr[j+1] = tmp
      }
    }
  }
}

bubbleSort(arr);
console.log(arr)

/**
 * 2. 实现一个函数，可以对 url 中的 query 部分做拆解，
 * 返回一个 key - value 形式的 object
 */
const url = 'http://sample.com/?a=1&e&b=2&c=xx&d#hash';

const getQueryObj = (url) => {
  var obj = {}

  var str = url.split("?")[1].split("#")[0];
  var arr = str.split("&");
  
  for(let i=0; i<arr.length; i++){
    let strArr = arr[i].split("=")
    if(strArr[1]){
      obj[strArr[0]] = strArr[1]
    }
  }
  return obj
}

console.log(getQueryObj(url)) 


lib/src/swiper.dart
import 'dart:async';

import 'package:flutter/material.dart';

class Swiper extends StatefulWidget {
  final List<String> images;
  final double height;
  final double width;

  ///点击回调
  final ValueChanged<int> onTap;

  /// 切换动画
  final Curve curve;

  ///是否循环播放
  final bool loop;

  /// 是否显示指示器
  final bool showIndicator;

  /// 是否自动轮播
  final bool autoplay;

  ///指示器激活的颜色  默认蓝色
  final Color activeColor;

  /// 指示器默认的颜色 默认白色
  final Color defaultColor;

  ///轮播图滚动方向 默认水平
  final Axis scrollDirection;

  ///图片展示样式
  final BoxFit imgFit;

  ///指示器的大小
  final Size indicatorSize;

  ///图片圆角大小
  final BorderRadius borderRadius;
  // final Widget Function(bool active) indicatorBuilder;

  ///自定义单个指示器
  final IndexedWidgetBuilder indicatorBuilder;

  /// 图片padding
  ///
  final EdgeInsetsGeometry imagMargin;

  /// 每一项可以自定义样式
  final Decoration itemdDecoration;

  Swiper(this.images,
      {this.height = 200,
      this.width,
      this.onTap,
      this.curve = Curves.linear,
      this.loop = true,
      this.activeColor = Colors.blue,
      this.defaultColor = Colors.white,
      this.showIndicator = true,
      this.autoplay = true,
      this.scrollDirection = Axis.horizontal,
      this.imgFit = BoxFit.cover,
      this.indicatorSize = const Size(8, 8),
      this.borderRadius,
      this.indicatorBuilder,
      this.imagMargin,
      this.itemdDecoration

      // this.indicatorBuilder = (active) => active ? Placeholder() : 'xx',
      })
      : assert(images != null);

  @override
  _SwiperState createState() => _SwiperState();
}

class _SwiperState extends State<Swiper> {
  PageController _pageController;
  int _currentIndex = 0;
  Timer _timer;

  @override
  void initState() {
    super.initState();
    _pageController = PageController(
      initialPage: _currentIndex,
    );
    _initTimer();
  }

  @override
  Widget build(BuildContext context) {
    return Stack(
      alignment: widget.scrollDirection == Axis.horizontal
          ? Alignment.bottomCenter
          : Alignment.centerRight,
      children: [
        _buildPageView(),
        _buildIndicator(),
      ],
    );
  }

  Widget _buildPageView() {
    var length = widget.images.length;
    return Container(
      height: widget.height,
      width: widget.width,
      child: PageView.builder(
        scrollDirection: widget.scrollDirection,
        controller: _pageController,
        onPageChanged: (index) {
          setState(() {
            _currentIndex = index;
          });
          _changePage();
        },
        itemBuilder: (context, index) {
          return GestureDetector(
            onPanDown: (details) {
              _cancelTimer();
            },
            onTap: () => widget.onTap(index % length),
            child: Container(
              margin: widget.imagMargin ?? EdgeInsets.all(0),
              child: ClipRRect(
                borderRadius: widget.borderRadius ?? BorderRadius.circular(0),
                child: Container(
                  decoration: widget.itemdDecoration ?? BoxDecoration(),
                  child: Image(
                    image: NetworkImage(widget.images[index % length]),
                    fit: widget.imgFit,
                  ),
                ),
              ),
            ),
          );
        },
      ),
    );
  }

  Widget _buildIndicator() {
    if (!widget.showIndicator) return SizedBox.shrink();

    if (widget.indicatorBuilder != null) {
      return widget.indicatorBuilder(
        context,
        _currentIndex % widget.images.length,
      );
    }

    return _buildDefaultIndicator();
  }

  Widget _buildDefaultIndicator() {
    var length = widget.images.length;
    var _width = widget.indicatorSize.width;
    var _height = widget.indicatorSize.height;

    return Positioned(
      bottom: widget.scrollDirection == Axis?.horizontal ? 10 : null,
      right: widget.scrollDirection == Axis?.vertical ? 10 : null,
      child: widget.scrollDirection == Axis.horizontal
          ? Container(
              child: Row(
                children: widget.images.map(
                  (e) {
                    return GestureDetector(
                        onTap: () => _ontap(e),
                        child: Padding(
                          padding: const EdgeInsets.symmetric(horizontal: 3),
                          child: ClipOval(
                            child: Container(
                              width: _width,
                              height: _height,
                              color: e == widget.images[_currentIndex % length]
                                  ? widget.activeColor
                                  : widget.defaultColor,
                            ),
                          ),
                        ));
                  },
                ).toList(),
              ),
            )
          : Container(
              child: Column(
                children: widget.images.map(
                  (e) {
                    return GestureDetector(
                      onTap: () => _ontap(e),
                      child: Padding(
                        padding: const EdgeInsets.symmetric(vertical: 3),
                        child: ClipOval(
                          child: Container(
                            width: _width,
                            height: _height,
                            color: e == widget.images[_currentIndex % length]
                                ? widget.activeColor
                                : widget.defaultColor,
                          ),
                        ),
                      ),
                    );
                  },
                ).toList(),
              ),
            ),
    );
  }

  _changePage() {
    Timer(Duration(milliseconds: 300), () {
      _pageController.jumpToPage(_currentIndex);
    });
  }

  // 点击图片回调
  _ontap(e) {
    var index = widget.images.indexOf(e);
    _currentIndex = index;
    goToIndex();
  }

  // 设置轮播定时器
  _initTimer() {
    if (_timer == null) {
      if (widget.autoplay != true) return;
      _timer = Timer.periodic(
        Duration(seconds: 3),
        (t) {
          var length = widget.images.length;
          if (widget.loop != true && (_currentIndex % length == length - 1)) {
            return _timer = null;
          }
          _currentIndex = (_currentIndex + 1) % length;
          goToIndex();
        },
      );
    }
  }

  // 滚动到指定页面
  void goToIndex() {
    _pageController.animateToPage(
      _currentIndex,
      duration: Duration(milliseconds: 300),
      curve: widget.curve,
    );
  }

  // 手动滑动的时候取消定时器
  _cancelTimer() {
    if (_timer != null) {
      _timer.cancel();
      _timer = null;
      _initTimer();
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:xhb_swiper/xhb_swiper.dart';

void main() => runApp(MyApp());

class MyWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) => ErrorWidget("flutter 自定义");
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<String> _listes = [
    "https://yy0.tv.cmvideo.cn:8443/migutv/res/2019/10/17/59L11DV8GKF0.jpg",
    "https://yy0.tv.cmvideo.cn:8443/migutv/res/2019/09/30/59IHTLCINHHH.jpg",
    "https://yy1.tv.cmvideo.cn:8443/migutv-clt/pomsimage/1504/026/730/202003240715_CCTVNEWS_1500000_20200324_35370063_0_327_HSJ1080H.jpg"
  ];
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Column(
        children: [
          Swiper(
            _listes,
            onTap: (value) {
              print(value);
            },
            imagMargin: EdgeInsets.all(20),
            // indicatorBuilder: (context, index) {
            //   return Container(
            //     width: 100,
            //     height: 30,
            //     alignment: Alignment.center,
            //     color: Colors.white,
            //     child: Text(
            //       '$index/3',
            //       style: TextStyle(
            //         color: Colors.red,
            //         fontWeight: FontWeight.w700,
            //         fontSize: 20,
            //       ),
            //     ),
            //   );
            // },
            scrollDirection: Axis.horizontal,
            defaultColor: Colors.red,
            loop: true,
            autoplay: true,
            activeColor: Colors.yellow,
            indicatorSize: Size(10, 10),
            borderRadius: BorderRadius.circular(20),
          ),
          Swiper(
            _listes,
            onTap: (value) {
              print(value);
            },
            indicatorBuilder: (context, index) {
              return Container(
                width: 100,
                height: 30,
                alignment: Alignment.center,
                color: Colors.white,
                child: Text(
                  '${index + 1}/3',
                  style: TextStyle(
                    color: Colors.red,
                    fontWeight: FontWeight.w700,
                    fontSize: 20,
                  ),
                ),
              );
            },
          ),
          Swiper(
            _listes,
            borderRadius: BorderRadius.circular(10),
            onTap: (value) {
              print(value);
            },
            scrollDirection: Axis.vertical,
            imagMargin: EdgeInsets.all(30),
            itemdDecoration: BoxDecoration(
              border: Border.all(color: Colors.yellow, width: 10),
            ),
            defaultColor: Colors.red,
          ),
        ],
      ),
    );
  }
}
